File: modules/drivers/sensors/bmp280/bmp280.js
/*
 * Copyright (c) 2016-2023  Moddable Tech, Inc.
 *
 *   This file is part of the Moddable SDK Runtime.
 *
 *   The Moddable SDK Runtime is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU Lesser General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 *
 *   The Moddable SDK Runtime is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU Lesser General Public License for more details.
 *
 *   You should have received a copy of the GNU Lesser General Public License
 *   along with the Moddable SDK Runtime.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
/*
    BMP280 - temp/pressure
	Datasheet - https://www.bosch-sensortec.com/media/boschsensortec/downloads/datasheets/bst-bmp280-ds001.pdf
    Implementation based on Adafruit https://github.com/adafruit/Adafruit_BMP280_Library
*/
import Timer from "timer";

const Register = Object.freeze({
	BMP280_SOFTRESET: 0xE0,
	BMP280_STATUS: 0xF3,
	BMP280_CONTROL: 0xF4,
	BMP280_CONFIG: 0xF5,
	BMP280_PRESSUREDATA: 0xF7,
	BMP280_TEMPDATA: 0xFA,
	BMP280_CHIPID: 0xD0
});

const Config = Object.freeze({
	Sampling: {
		NONE: 0x00,
		X1: 0x01,
		X2: 0x02,
		X4: 0x03,
		X8: 0x04,
		X16: 0x05
	},
	Filter: {
		OFF: 0x00,
		X2: 0x01,
		X4: 0x02,
		X8: 0x03,
		X16: 0x04
	},
	Mode: {
		SLEEP: 0x00,
		FORCED: 0x01,
		NORMAL: 0x03,
		SOFT_RESET: 0xB6
	},
	Standby: {
		MS_1:	0x00,
		MS_63:	0x01,
		MS_125:	0x02,
		MS_250:	0x03,
		MS_500: 0x04,
		MS_1000: 0x05,
		MS_2000: 0x06,
		MS_4000: 0x07
	}
}, true);

class aHostObject @ "xs_bmp280_host_object_destructor" {
	constructor() @ "xs_bmp280_host_object_constructor";
}

class BMP280 extends aHostObject {
	#io;
	#byteBuffer;
	#wordBuffer;
	#valueBuffer;

	constructor(options) {
		super(options);
		const io = this.#io = new options.sensor.io({
			hz: 400_000,
			address: 0x76,
			...options.sensor
		});

		const bBuf = this.#byteBuffer = new Uint8Array(1);
		const wBuf = this.#wordBuffer = new Uint8Array(2);
		this.#valueBuffer = new Uint8Array(6);

		bBuf[0] = Register.BMP280_CHIPID;
		io.write(bBuf);
		io.read(bBuf);
		if (0x58 !== bBuf[0]) {
			this.close();
			throw new Error("unexpected sensor");
		}

		wBuf[0] = Register.BMP280_SOFTRESET;
		wBuf[1] = Config.Mode.SOFT_RESET;
		io.write(wBuf);
		Timer.delay(10);

		this.#initialize();

		this.configure({
			mode: Config.Mode.NORMAL,
			temperatureSampling: Config.Sampling.X2,
			pressureSampling: Config.Sampling.X16,
			filter: Config.Filter.X16,
			standbyDuration: Config.Standby.MS_500
		});
	}
	configure(options) {
		const io = this.#io;
		const bBuf = this.#byteBuffer;
		const wBuf = this.#wordBuffer;

		bBuf[0] = Register.BMP280_CONFIG;
		io.write(bBuf);
		io.read(bBuf);
		let configReg = bBuf[0];

		bBuf[0] = Register.BMP280_CONTROL;
		io.write(bBuf);
		io.read(bBuf);
		let measureReg = bBuf[0];

		if (undefined != options.mode)
			measureReg = (measureReg & 0b1111_1100) | options.mode;
		if (undefined != options.pressureSampling)
			measureReg = (measureReg & 0b1110_0011) | (options.pressureSampling << 2);
		if (undefined != options.temperatureSampling)
			measureReg = (measureReg & 0b0001_1111) | (options.temperatureSampling << 5);
		if (undefined != options.standbyDuration)
			configReg = (configReg & 0b0001_1111) | (options.standbyDuration << 5);
		if (undefined != options.filter)
			configReg = (configReg & 0b1110_0011) | (options.filter << 2);


		wBuf[0] = Register.BMP280_CONFIG;
		wBuf[1] = configReg;
		io.write(wBuf);

		wBuf[0] = Register.BMP280_CONTROL;
		wBuf[1] = measureReg;
		io.write(wBuf);
	}
	sample() {
		const io = this.#io;
		const bBuf = this.#byteBuffer;
		const vBuf = this.#valueBuffer;
		
		bBuf[0] = Register.BMP280_PRESSUREDATA;
		io.write(bBuf);
		io.read(vBuf);
		const rawP = (vBuf[0] << 16) | (vBuf[1] << 8) | vBuf[2];
		const rawT = (vBuf[3] << 16) | (vBuf[4] << 8) | vBuf[5];
		const val = this.#calculate(rawT, rawP);
		return { barometer: { pressure: val.pressure }, thermometer: { temperature: val.temperature } };
	}
	#close() @ "xs_bmp280_close";
	close() {
		this.#close();
		this.#io?.close();
		this.#io = undefined;
	}	
	#calculate(rawTemp, rawPressure) @ "xs_bmp280_calculate";
	#setCalibration(calibrate) @ "xs_bmp280_setCalibration";
	#initialize() {
		let calib = {};
		calib.T1 = this.#read16LE(0x88);
		calib.T2 = this.#readS16LE(0x8A);
		calib.T3 = this.#readS16LE(0x8C);

		calib.P1 = this.#read16LE(0x8E);
		calib.P2 = this.#readS16LE(0x90);
		calib.P3 = this.#readS16LE(0x92);
		calib.P4 = this.#readS16LE(0x94);
		calib.P5 = this.#readS16LE(0x96);
		calib.P6 = this.#readS16LE(0x98);
		calib.P7 = this.#readS16LE(0x9A);
		calib.P8 = this.#readS16LE(0x9C);
		calib.P9 = this.#readS16LE(0x9E);

		this.#setCalibration(calib);
	}
	#twoC16(val) {
		return (val > 32767) ? -(65535 - val + 1) : val;
	}
	#read16(reg) {
		const io = this.#io;
		const bBuf = this.#byteBuffer;
		const wBuf = this.#wordBuffer;
		bBuf[0] = reg;
		io.write(bBuf);
		io.read(wBuf);
		return (wBuf[0] << 8) | wBuf[1];
	}
	#readS16(reg) {
		return this.#twoC16(this.#read16(reg));
	}
	#read16LE(reg) {
		let v = this.#read16(reg);
		return ((v & 0xff00) >> 8) | ((v & 0xff) << 8);
	}
	#readS16LE(reg) {
		return this.#twoC16(this.#read16LE(reg));
	}
}

export { BMP280 as default, BMP280, Config };

File: modules/drivers/sensors/bmp280/bmp280.c
/*
 * Copyright (c) 2016-2021  Moddable Tech, Inc.
 *
 *   This file is part of the Moddable SDK Runtime.
 *
 *   The Moddable SDK Runtime is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU Lesser General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 *
 *   The Moddable SDK Runtime is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU Lesser General Public License for more details.
 *
 *   You should have received a copy of the GNU Lesser General Public License
 *   along with the Moddable SDK Runtime.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
/*
    BMP280 - temp/pressure
    Implementation based on Adafruit https://github.com/adafruit/Adafruit_BMP280_Library
*/
#include "xsHost.h"
#include "xsmc.h"
#include "mc.xs.h"

typedef struct calibrationSet {
	xsSlot		obj;
	xsMachine	*the;

	uint16_t	T1;
	int16_t		T2;
	int16_t		T3;

	uint16_t	P1;
	int16_t		P2;
	int16_t		P3;
	int16_t		P4;
	int16_t		P5;
	int16_t		P6;
	int16_t		P7;
	int16_t		P8;
	int16_t		P9;
} calibrationSet, *calibrationPtr;

void xs_bmp280_host_object_constructor(xsMachine *the)
{
	calibrationPtr calib;

	calib = c_malloc(sizeof(calibrationSet));
	if (!calib)
		xsUnknownError("no memory");

	xsmcSetHostData(xsThis, calib);
	calib->obj = xsThis;
	calib->the = the;
	xsRemember(xsThis);
}

void xs_bmp280_host_object_destructor(void *data)
{
	if (NULL == data)
		return;

	c_free(data);
}

void xs_bmp280_setCalibration(xsMachine *the)
{
	calibrationPtr calib = (calibrationPtr)xsmcGetHostData(xsThis);

	xsmcVars(1);

	xsmcGet(xsVar(0), xsArg(0), xsID_T1);
	calib->T1 = xsmcToInteger(xsVar(0));
	xsmcGet(xsVar(0), xsArg(0), xsID_T2);
	calib->T2 = xsmcToInteger(xsVar(0));
	xsmcGet(xsVar(0), xsArg(0), xsID_T3);
	calib->T3 = xsmcToInteger(xsVar(0));
	xsmcGet(xsVar(0), xsArg(0), xsID_P1);
	calib->P1 = xsmcToInteger(xsVar(0));
	xsmcGet(xsVar(0), xsArg(0), xsID_P2);
	calib->P2 = xsmcToInteger(xsVar(0));
	xsmcGet(xsVar(0), xsArg(0), xsID_P3);
	calib->P3 = xsmcToInteger(xsVar(0));
	xsmcGet(xsVar(0), xsArg(0), xsID_P4);
	calib->P4 = xsmcToInteger(xsVar(0));
	xsmcGet(xsVar(0), xsArg(0), xsID_P5);
	calib->P5 = xsmcToInteger(xsVar(0));
	xsmcGet(xsVar(0), xsArg(0), xsID_P6);
	calib->P6 = xsmcToInteger(xsVar(0));
	xsmcGet(xsVar(0), xsArg(0), xsID_P7);
	calib->P7 = xsmcToInteger(xsVar(0));
	xsmcGet(xsVar(0), xsArg(0), xsID_P8);
	calib->P8 = xsmcToInteger(xsVar(0));
	xsmcGet(xsVar(0), xsArg(0), xsID_P9);
	calib->P9 = xsmcToInteger(xsVar(0));
}

void xs_bmp280_close(xsMachine *the)
{
	calibrationPtr calib = xsmcGetHostData(xsThis);
	if (NULL == calib)
		return;

	xsForget(calib->obj);
	xsmcSetHostData(xsThis, NULL);
	c_free(calib);
}

void xs_bmp280_calculate(xsMachine *the)
{
	calibrationPtr calib = xsmcGetHostData(xsThis);
	if (NULL == calib)
		return;

	int32_t var1, var2;
	int32_t adc_T = xsmcToInteger(xsArg(0));
	int32_t	t_fine;

	adc_T >>= 4;
	var1 = ((((adc_T >> 3) - ((int32_t)calib->T1 << 1))) *
		((int32_t)calib->T2)) >> 11;

	var2 = (((((adc_T >> 4) - ((int32_t)calib->T1)) *
			((adc_T >> 4) - ((int32_t)calib->T1))) >> 12) *
			((int32_t)calib->T3)) >> 14;

	t_fine = var1 + var2;

	double T = ((t_fine * 5 + 128) >> 8);
	T /= 100.0;

	xsResult = xsNewHostObject(c_free);
	xsmcVars(1);
	xsVar(0) = xsNumber(T);
	xsmcSet(xsResult, xsID_temperature, xsVar(0));

	int64_t pvar1, pvar2, p;

	int32_t adc_P = xsmcToInteger(xsArg(1));
	adc_P >>= 4;

	pvar1 = ((int64_t)t_fine) - 128000;
	pvar2 = pvar1 * pvar1 * (int64_t)calib->P6;
	pvar2 = pvar2 + ((pvar1 * (int64_t)calib->P5) << 17);
	pvar2 = pvar2 + (((int64_t)calib->P4) << 35);
	pvar1 = ((pvar1 * pvar1 * (int64_t)calib->P3) >> 8) +
			((pvar1 * (int64_t)calib->P2) << 12);
	pvar1 = (((((int64_t)1) << 47) + pvar1)) * ((int64_t)calib->P1) >> 33;

	if (pvar1 == 0)
		xsUnknownError("bmp280 div by zero");

	p = 1048576 - adc_P;
	p = (((p << 31) - pvar2) * 3125) / pvar1;
	pvar1 = (((int64_t)calib->P9) * (p >> 13) * (p >> 13)) >> 25;
	pvar2 = (((int64_t)calib->P8) * p) >> 19;

	p = ((p + pvar1 + pvar2) >> 8) + (((int64_t)calib->P7) << 4);

	double P = (double)p / 256.0;

	xsVar(0) = xsNumber(P);
	xsmcSet(xsResult, xsID_pressure, xsVar(0));
}

